lib/searcher/src/main/scala/org/enso/searcher/sql/SqlSuggestionsRepo.scala

```diff
@@ -16,38 +17,139 @@ final class SqlSuggestionsRepo(implicit ec: ExecutionContext)
-    query.result.map(coll => joinedToSuggestion(coll).headOption)
+    query.result.map(coll => joinedToSuggestions(coll).headOption)
   }
 
-  /** @inheritdoc **/
-  override def insert(suggestion: Suggestion): DBIO[Long] = {
+  /** @inheritdoc */
+  override def insert(suggestion: Suggestion): DBIO[Option[Long]] = {
     val (suggestionRow, args) = toSuggestionRow(suggestion)
-    for {
-      id <- suggestions.returning(suggestions.map(_.id)) += suggestionRow
-      _  <- arguments ++= args.map(toArgumentRow(id, _))
+    val query = for {
+      id <- Suggestions.returning(Suggestions.map(_.id)) += suggestionRow
+      _ <- Arguments ++= args.zipWithIndex.map {
+          case (argument, ix) => toArgumentRow(id, ix, argument)
+        }
+      _ <- incrementVersion
     } yield id
+    query.transactionally.asTry.map {
+      case Failure(_)  => None
+      case Success(id) => Some(id)
+    }
+  }
+
+  /** @inheritdoc */
+  override def insertAll(
+    suggestions: Seq[Suggestion]
+  ): DBIO[Seq[Option[Long]]] = {
+    DBIO.sequence(suggestions.map(insert))
+  }
+
+  /** @inheritdoc */
+  override def remove(suggestion: Suggestion): DBIO[Option[Long]] = {
+    val (raw, _) = toSuggestionRow(suggestion)
+    val selectQuery = Suggestions
+      .filter(_.kind === raw.kind)
+      .filter(_.name === raw.name)
+      .filter(_.scopeStart === raw.scopeStart)
+      .filter(_.scopeEnd === raw.scopeEnd)
+    val deleteQuery = for {
+      rows <- selectQuery.result
+      n    <- selectQuery.delete
+      _    <- if (n > 0) incrementVersion else DBIO.successful(())
+    } yield rows.flatMap(_.id).headOption
+    deleteQuery.transactionally
+  }
+
+  /** @inheritdoc */
+  override def removeAll(
+    suggestions: Seq[Suggestion]
+  ): DBIO[Seq[Option[Long]]] = {
+    DBIO.sequence(suggestions.map(remove))
+  }
+
+  /** @inheritdoc */
+  override def currentVersion: DBIO[Long] = {
+    for {
+      versionOpt <- Versions.result.headOption
+    } yield versionOpt.flatMap(_.id).getOrElse(0L)
+  }
+
+  /** Increment the current version of the repo. */
+  private def incrementVersion: DBIO[Long] = {
+    val increment = for {
+      version <- Versions.returning(Versions.map(_.id)) += VersionRow(None)
+      _       <- Versions.filterNot(_.id === version).delete
+    } yield version
+    increment.transactionally
+  }
+
+  /** Create a search query by the provided parameters. */
+  private def searchQuery(
+    selfType: Option[String],
+    returnType: Option[String],
+    kinds: Option[Seq[Suggestion.Kind]]
+  ): Query[
+    (Rep[Option[ArgumentsTable]], SuggestionsTable),
+    (Option[ArgumentRow], SuggestionRow),
+    Seq
+  ] = {
+    joined
+      .filterOpt(selfType) {
+        case (row, value) => row._2.selfType === value
+      }
+      .filterOpt(returnType) {
+        case (row, value) => row._2.returnType === value
+      }
+      .filterOpt(kinds) {
+        case (row, value) => row._2.kind inSet value.map(SuggestionKind(_))
+      }
```

Comment:
It is a potential bottleneck. It joins two tables and next performs a full table scan.

Reply:
Good point. I spent some time today, checking the query plans, adding logging and missing indexes. All of the search queries make use of indexes now.

Also, search query returns only `suggestions.id` field, and we don't need the joined query at all here. End up removing the join.

Below are the resulting SQL queries that are being executed:

```sql
explain query plan select "id" from "suggestions" where "return_type" = 'MyType';
QUERY PLAN
`--SEARCH TABLE suggestions USING COVERING INDEX suggestions_return_type_idx (return_type=?)
```

```sql
explain query plan select "id" from "suggestions" where "self_type" = 'Main';
QUERY PLAN
`--SEARCH TABLE suggestions USING COVERING INDEX suggestions_self_type_idx (self_type=?)
```

```sql
explain query plan select "id" from "suggestions" where "kind" in (0, 3);
QUERY PLAN
`--SEARCH TABLE suggestions USING COVERING INDEX suggestions_kind_idx (kind=?)
```